### PR TITLE
Tab SSO C# to support authentication in browser.

### PR DESCRIPTION
**Pull request by HunaidHanfee-MSFT**

### CODE REVIEW

File: samples/tab-sso/csharp/TeamsTabSSO/wwwroot/js/msal-auth.js [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                if (error.errorMessage.indexOf("interaction_required") >= 0) {
+                    return this.app
+                        .acquireTokenPopup({ scopes })
+                        .then((authResponse) => authResponse.accessToken);
+                } else {
+                    return Promise.reject(error);
+                }
+            });
+    }
+
```
Comment (ChetanSharma-msft) [inline samples/tab-sso/csharp/TeamsTabSSO/wwwroot/js/msal-auth.js:82]: Remove extra empty lines.

File: samples/tab-sso/csharp/TeamsTabSSO/wwwroot/js/msal-auth.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+    getUserInfo(principalName) {
+        this.getToken().then((token) => {
+            if (principalName) {
+                let graphUrl = "https://graph.microsoft.com/v1.0/users/" + principalName;
+                $.ajax({
+                    url: graphUrl,
```
Comment (ChetanSharma-msft) [inline samples/tab-sso/csharp/TeamsTabSSO/wwwroot/js/msal-auth.js:89]: Add proper spacing and line break.

File: samples/tab-sso/csharp/TeamsTabSSO/wwwroot/js/msal-auth.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    getUser() {
+        let activeAccount = this.app.getActiveAccount();
+        if (!activeAccount) {
+            const allAccounts = this.app.getAllAccounts();
+            if (allAccounts.length === 1) {
+                this.app.setActiveAccount(allAccounts[0]);
+                activeAccount = allAccounts[0];
+            }
+        }
+        return Promise.resolve(activeAccount);
```
Comment (ChetanSharma-msft) [inline samples/tab-sso/csharp/TeamsTabSSO/wwwroot/js/msal-auth.js:64]: Add proper spacing and line break.

File: samples/tab-sso/csharp/TeamsTabSSO/wwwroot/js/auth.js [hunk 4]
```diff
@@ -1,35 +1,41 @@
...
+            })
+            .catch((error) => {
+                console.log(error);
+                if (error === "invalid_grant") {
+                    // Display in-line button so user can consent
+                    $("#divError").text("Error while exchanging for Server token - invalid_grant - User or admin consent is required.");
+                    $("#divError").show();
+                    $("#consent").show();
+                } else {
+                    // Something else went wrong
```
Comment (ChetanSharma-msft) [inline samples/tab-sso/csharp/TeamsTabSSO/wwwroot/js/auth.js:19]: Can we write some console.log or message?

**Review by ChetanSharma-msft — CHANGES_REQUESTED** [review]
Reviewed and added my comments.

File: samples/tab-sso/csharp/TeamsTabSSO/Views/Home/Index.cshtml [hunk 5]
```diff
@@ -1,7 +1,96 @@
...
+    </style>
+
+    <script type="text/javascript" src="https://alcdn.msauth.net/browser/2.24.0/js/msal-browser.min.js"></script>
+    <script>
+        const authService = new MsalAuthService("@ViewBag.clientId", "@ViewBag.applicationIdURI");
+        $(document).ready(function() {
+            authService
+                .isCallback()
+                .then((isCallback) => {
+                    if (!isCallback) {
```
[the remark this answers is not in the document]
Comment (HunaidHanfee-MSFT) [reply samples/tab-sso/csharp/TeamsTabSSO/Views/Home/Index.cshtml:31]: Done

File: samples/tab-sso/csharp/TeamsTabSSO/wwwroot/js/msal-auth.js [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                    type: "GET",
+                    beforeSend: function (request) {
+                        request.setRequestHeader("Authorization", `Bearer ${token}`);
+                    },
+                    success: function (profile) {
+                        let profileDiv = $("#divGraphProfile");
+                        profileDiv.empty();
+                        //$("<div>")
+                        //    .append($("<pre>").text(JSON.stringify(profile, null, "\t")))
+                        //    .appendTo(profileDiv);
```
[the remark this answers is not in the document]
Comment (HunaidHanfee-MSFT) [reply samples/tab-sso/csharp/TeamsTabSSO/wwwroot/js/msal-auth.js]: Done

File: samples/tab-sso/csharp/TeamsTabSSO/wwwroot/js/msal-auth.js (hunk 3 above)
Comment (HunaidHanfee-MSFT) [reply samples/tab-sso/csharp/TeamsTabSSO/wwwroot/js/msal-auth.js:66]: Done

File: samples/tab-sso/csharp/TeamsTabSSO/Views/Home/BrowserRedirect.cshtml [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+<head>
+    <script type="text/javascript" src="https://alcdn.msauth.net/browser/2.24.0/js/msal-browser.min.js"></script>
+
+    <script>
+        const authService = new MsalAuthService("@ViewBag.clientId", "@ViewBag.applicationIdURI");
+        $(document).ready(function() {
+            authService
+                .isCallback()
+                .then((isCallback) => {
+                    if (!isCallback) {
```
[the remark this answers is not in the document]
Comment (HunaidHanfee-MSFT) [reply samples/tab-sso/csharp/TeamsTabSSO/Views/Home/BrowserRedirect.cshtml:16]: Done

File: samples/tab-sso/csharp/README.md [hunk 8]
```diff
@@ -17,14 +17,23 @@ This sample illustrates how to implement SSO authentication for Teams Tab.
...
 ![Config](Images/config_teams.png)
 
 ![Config SSO](Images/config_sso_teams.png)
 
+### Browser SSO
+To test this feature, open your App base URI or Ngrok URI on browser.
+
+URI should be like `https://%ngrokDomain%.ngrok.io/Home/`
+
+![Browser SSO](Images/browser_sso.png)
```
Comment (Wajeed-msft) [inline samples/tab-sso/csharp/README.md:33]: Could you please change this to "Login Outside Teams Apps"? Also, rename the image.

**Review by Wajeed-msft — APPROVED** [review]

File: samples/tab-sso/csharp/README.md (hunk 8 above)
Comment (HunaidHanfee-MSFT) [reply samples/tab-sso/csharp/README.md]: Done

**Review by ChetanSharma-msft — APPROVED** [review]
Changes looks fine, approving it.

**Pull request merged**